In [54]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression

In [55]:
X , y = make_regression(n_samples=1000 , n_features=3 , noise=10 , random_state=42)

In [56]:
print(X.shape)

(1000, 3)


In [57]:
print(y.shape)

(1000,)


In [58]:
X = np.c_[np.ones(X.shape[0]), X]

In [59]:
y = y.reshape(len(y),1)

In [60]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [61]:
def calculate_loss(X, y, weights):
    
# # Example data
# X = np.array([[1, 2], [3, 4], [5, 6]])  
# y = np.array([5, 11, 17])              
# weights = np.array([1, 1])             

    #  Calculate predictions by multiplying inputs (X) with weights:)
    predictions = X.dot(weights)    #predictions=[1⋅1+2⋅1,3⋅1+4⋅1,5⋅1+6⋅1]=[3,7,11]
    
    #  Calculate the errors (difference between predictions and actual values):)
    errors = predictions - y  #errors=[3−5,7−11,11−17]=[−2,−4,−6]
    
    # Calculate Mean Squared Error (MSE):)
    mse = np.mean(errors ** 2)   #mse=error **2/the number of samples  
    
    #  Divide by 2 to match the gradient calculation format:)
    loss = mse / 2  
    
    return loss

Gradient Descent

In [62]:
def gradient_descent(X, y, num_iterations , freq_print):
    alpha = 0.001
    W = np.random.randn(X.shape[1], 1)
    for i in range(num_iterations) :    
        yhat = X.dot(W)
        error = yhat - y
        gradient = X.T.dot(error)
        W = W - alpha * gradient / len(X)
        if i % freq_print == 0:
            print(calculate_loss(X , y , W))
        
        
    return W             

In [63]:
W = gradient_descent(X_test , y_test , 10000 , 1000)

8623.773822508063
1258.645602129646
234.32705135837605
79.73275675928022
54.50717756389231
50.09732091589181
49.281976727681034
49.12472572267961
49.093475920034834
49.087138381050046


In [64]:
W 

array([[ 1.60789256],
       [98.19382595],
       [81.98366123],
       [25.92512203]])

To test the gradient descent

this function give the most efficient weight and I print the loss for this weight

In [65]:
def least_weight(X, y):
    X_transpose = X.T
    W = np.linalg.inv(X_transpose @ X) @ X_transpose @ y
    
    
    return W
    

print the output for the test data withe the most efficient weights

In [66]:
print(calculate_loss(X_test,y_test,least_weight(X_test,y_test)))

49.08549457182744


The gradient descent final loss output is almost equal to the least loss output.......................
                                   49.087106110189694  |||  49.08549457182744

Predicting the output for the test data

In [67]:
def predict(W, X):
    return X.dot(W)

In [68]:
predictions = predict(W, X_test)

In [69]:
comp = np.concatenate((predictions.reshape(len(predictions),1),y_test),axis=1)

In [70]:
comp

array([[-7.48480516e+01, -8.22932793e+01],
       [ 1.12992998e+02,  1.04807640e+02],
       [ 5.11004336e+01,  5.51531708e+01],
       [ 1.80672430e+02,  1.75842100e+02],
       [-2.25489115e+01, -1.28221444e+01],
       [-1.62555884e+02, -1.58696775e+02],
       [ 1.34343389e+02,  1.17792000e+02],
       [ 6.92082751e+01,  6.45922450e+01],
       [ 2.31855558e+02,  2.36283381e+02],
       [-1.12751845e+02, -1.09113405e+02],
       [-1.06858838e+02, -1.03799700e+02],
       [ 1.18704144e+02,  1.14628724e+02],
       [ 3.13584244e+01,  3.15693700e+01],
       [-1.59157790e+01,  2.21697784e+00],
       [ 3.81891418e+01,  4.32591823e+01],
       [-4.93418277e+01, -4.58647575e+01],
       [-1.90626635e+02, -1.81070715e+02],
       [ 6.19596537e+01,  7.58102653e+01],
       [-7.78907615e+00, -6.02334948e+00],
       [ 4.37302086e+01,  2.62678736e+01],
       [ 1.02933391e+02,  1.01897247e+02],
       [ 3.86025771e+01,  4.45269699e+01],
       [-1.27620252e+02, -1.25496420e+02],
       [ 1.

Checking the performance of the model


In [71]:
def mean_squared_error(y, y_pred):
    mse = (y - y_pred) ** 2
    return np.mean(mse)


In [72]:
mean_squared_error(y_test, predictions)

np.float64(98.17167269487587)

In [73]:
def mean_absolute_error(y, y_pred):
    mae = np.abs(y - y_pred)
    return np.mean(mae)

In [74]:
mean_absolute_error(y_test, predictions)

np.float64(8.058155216434699)